compiler/src/dotty/tools/dotc/core/TypeComparer.scala

```diff
@@ -840,8 +843,11 @@ class TypeComparer(initctx: Context) extends ConstraintHandling[AbsentContext] w
                     gadtBoundsContain(tycon1sym, tycon2) ||
                     gadtBoundsContain(tycon2sym, tycon1)
                   ) &&
-                  isSubType(tycon1.prefix, tycon2.prefix) &&
-                  isSubArgs(args1, args2, tp1, tparams)
+                  isSubType(tycon1.prefix, tycon2.prefix) && {
+                    // check both tycons to deal with the case when they are equal b/c of GADT constraint
+                    val tyconIsInjective = tycon1sym.isClass || tycon2sym.isClass
+                    isSubArgs(args1, args2, tp1, tparams, inferGadtBounds = tyconIsInjective)
```

Comment:
Would it not be better if `frozenGadt` was set around isSubArgs instead of for each individual arg? Fewer state changes and simpler logic since no additional parameter needs to be passed to `isSubArgs`. 

The only thing that changes is that `frozenGadt` then also affects `compareCaptured`, but I don't see why it should not do that.

Reply:
Hm, I agree that the code will be a bit simpler if we just set `frozenGadt` around `isSubArgs`.

I believe that `frozenGadt` already affects `compareCaptured`, since it calls `isSubArg`.